File: Algorithms/Scala/RandomSearch/src/main/scala/OSOL/Extremum/Algorithms/Scala/RandomSearch.scala
package OSOL.Extremum.Algorithms.Scala

import OSOL.Extremum.Cores.JVM.Optimization._
import OSOL.Extremum.Cores.JVM.Optimization.Nodes.{GeneralNode, SetParametersNode, TerminationViaMaxTime}
import OSOL.Extremum.Cores.JVM.Random.GoRN
import OSOL.Extremum.Cores.JVM.Vectors.RealVector
import OSOL.Extremum.Cores.JVM.Vectors.RealVector.Converters._

object RandomSearch {

  private val currentPointName = "currentPoint"
  private val currentPointEfficiencyName = "currentPointEfficiency"
  private val radiusParameterName = "r"

  private def generateRandomInSphere(currentPoint: RealVector, radius: Double, area: Area): RealVector = {
    val normallyDistributed = GoRN.getNormal(area.mapValues(_ => (0.0, 1.0)))
    val r = math.sqrt(normallyDistributed.values.map(v => v * v).sum)
    (currentPoint + normallyDistributed * (GoRN.getContinuousUniform(-1.0, 1.0) / r)).constrain(area)
  }

  private final class GenerateInitialPointNode(override val nodeId: java.lang.Integer) extends GeneralNode[RealVector, java.lang.Double, RealVector](nodeId) {

    override def initialize(f: Map[String, java.lang.Double] => java.lang.Double, area: Area, state: State[RealVector, java.lang.Double, RealVector]): Unit = {
      val initialPoint: RealVector = GoRN.getContinuousUniform(area)
      state.setParameter(currentPointName, initialPoint)
      state.setParameter(currentPointEfficiencyName, initialPoint.getPerformance(f))
    }

    override def process(f: Map[String, java.lang.Double] => java.lang.Double, area: Area, state: State[RealVector, java.lang.Double, RealVector]): Unit = { }

  }

  private final class SampleNewPointNode_FixedStep(override val nodeId: java.lang.Integer) extends GeneralNode[RealVector, java.lang.Double, RealVector](nodeId) {

    override def initialize(f: Map[String, java.lang.Double] => java.lang.Double, area: Area, state: State[RealVector, java.lang.Double, RealVector]): Unit = { }

    override def process(f: Map[String, java.lang.Double] => java.lang.Double, area: Area, state: State[RealVector, java.lang.Double, RealVector]): Unit = {
      val currentPoint = state.getParameter[RealVector](currentPointName)
      val currentPointEfficiency = state.getParameter[Double](currentPointEfficiencyName)
      val r = state.getParameter[Double](radiusParameterName)

      val newPoint = generateRandomInSphere(currentPoint, r, area)
      val newPointEfficiency = newPoint.getPerformance(f)

      if (newPointEfficiency < currentPointEfficiency) {
        state.setParameter(currentPointName, newPoint)
        state.setParameter(currentPointEfficiencyName, newPointEfficiency)
      }
    }

  }

  private final class SetBestNode(override val nodeId: java.lang.Integer) extends GeneralNode[RealVector, java.lang.Double, RealVector](nodeId) {

    override def initialize(f: Map[String, java.lang.Double] => java.lang.Double, area: Area, state: State[RealVector, java.lang.Double, RealVector]): Unit = { }

    override def process(f: Map[String, java.lang.Double] => java.lang.Double, area: Area, state: State[RealVector, java.lang.Double, RealVector]): Unit = {
      state.result = Some(state.getParameter[RealVector](currentPointName))
    }

  }

  def createFixedStepRandomSearch(radius: java.lang.Double, maxTime: java.lang.Double): Algorithm[RealVector, java.lang.Double, RealVector] = {
    val FixedStep_nodes = Seq(
      new SetParametersNode[RealVector, java.lang.Double, RealVector](nodeId = 0, parameters = Map(radiusParameterName -> radius)),
      new GenerateInitialPointNode(nodeId = 1),
      new TerminationViaMaxTime[RealVector, java.lang.Double, RealVector](nodeId = 2, maxTime),
      new SampleNewPointNode_FixedStep(nodeId = 3),
      new SetBestNode(nodeId = 4)
    )

    val FixedStep_transitionMatrix: Seq[(java.lang.Integer, Option[java.lang.Integer], java.lang.Integer)] = Seq(
      (0, None, 1),
      (1, None, 2),
      (2, Some(0), 3),
      (2, Some(1), 4),
      (3, None, 2)
    )

    new Algorithm[RealVector, java.lang.Double, RealVector](nodes = FixedStep_nodes, transitionMatrix = FixedStep_transitionMatrix)
  }

}

File: Algorithms/Java/RandomSearch/src/main/java/OSOL/Extremum/Algorithms/Java/RandomSearch.java
package OSOL.Extremum.Algorithms.Java;

import OSOL.Extremum.Cores.JVM.Optimization.*;
import OSOL.Extremum.Cores.JVM.Optimization.Nodes.*;
import OSOL.Extremum.Cores.JVM.Random.GoRN;
import OSOL.Extremum.Cores.JVM.Vectors.RealVector;

import scala.Function1;
import scala.Option;
import scala.Tuple2;
import scala.Tuple3;
import scala.collection.immutable.Map;
import scala.collection.JavaConverters;
import scala.collection.immutable.Map$;

import java.lang.*;
import java.util.*;

public class RandomSearch {

    private static String currentPointName = "currentPoint";
    private static String currentPointEfficiencyName = "currentPointEfficiency";
    private static String radiusParameterName = "r";

    private static RealVector generateRandomInSphere(RealVector currentPoint, Double radius, Map<String, Tuple2<Double, Double>> area) {
        RealVector normallyDistributed = RealVector.Converters.IterableToRealVector(GoRN.getNormal(area.mapValues(a -> Tuple2.apply(0.0, 1.0))));
        int dim = currentPoint.elements().size();
        List<Double> values = JavaConverters.seqAsJavaList(normallyDistributed.elements().values().toSeq());

        double r = 0.0;
        for (Double v : values)
            r += v.doubleValue() * v;

        Double moveCoefficient = (GoRN.getContinuousUniform(-1.0, 1.0) / r);

        return (currentPoint.add(normallyDistributed.multiply(moveCoefficient))).constrain(area);
    }

    public static final class GenerateInitialPointNode extends GeneralNode<RealVector, Double, RealVector> {
        @Override
        public void initialize(Function1<Map<String, Double>, Double> f, Map<String, Tuple2<Double, Double>> area, State<RealVector, Double, RealVector> state) {
            RealVector initialPoint = RealVector.Converters.IterableToRealVector(GoRN.getContinuousUniform(area));
            state.setParameter(currentPointName, initialPoint);
            state.setParameter(currentPointEfficiencyName, initialPoint.getPerformance(f));
        }

        @Override
        public void process(Function1<Map<String, Double>, Double> f, Map<String, Tuple2<Double, Double>> area, State<RealVector, Double, RealVector> state) {

        }

        public GenerateInitialPointNode(Integer nodeId) {
            super(nodeId);
        }
    }

    public static final class SampleNewPointNode_FixedStep extends GeneralNode<RealVector, Double, RealVector> {
        @Override
        public void initialize(Function1<Map<String, Double>, Double> f, Map<String, Tuple2<Double, Double>> area, State<RealVector, Double, RealVector> state) {

        }

        @Override
        public void process(Function1<Map<String, Double>, Double> f, Map<String, Tuple2<Double, Double>> area, State<RealVector, Double, RealVector> state) {
            RealVector currentPoint = state.getParameter(currentPointName);
            Double currentPointEfficiency = state.getParameter(currentPointEfficiencyName);
            Double r = state.getParameter(radiusParameterName);

            RealVector newPoint = generateRandomInSphere(currentPoint, r, area);
            Double newPointEfficiency = newPoint.getPerformance(f);

            if (newPointEfficiency < currentPointEfficiency) {
                state.setParameter(currentPointName, newPoint);
                state.setParameter(currentPointEfficiencyName, newPointEfficiency);
            }
        }

        public SampleNewPointNode_FixedStep(Integer nodeId) {
            super(nodeId);
        }
    }

    public static final class SetBestNode extends GeneralNode<RealVector, Double, RealVector> {
        @Override
        public void initialize(Function1<Map<String, Double>, Double> f, Map<String, Tuple2<Double, Double>> area, State<RealVector, Double, RealVector> state) {

        }

        @Override
        public void process(Function1<Map<String, Double>, Double> f, Map<String, Tuple2<Double, Double>> area, State<RealVector, Double, RealVector> state) {
            RealVector currentPoint = state.getParameter(currentPointName);
            state.result_$eq(Option.apply(currentPoint));
        }

        public SetBestNode(Integer nodeId) {
            super(nodeId);
        }
    }

    public static Algorithm<RealVector, Double, RealVector> createFixedStepRandomSearch(Double radius, Double maxTime) {
        Map<String, Object> parameters = Map$.MODULE$.<String, Object>empty().$plus(Tuple2.apply(radiusParameterName, radius));

        List<GeneralNode<RealVector, Double, RealVector>> FixedStep_nodes = new ArrayList<>();
        FixedStep_nodes.add(new SetParametersNode<>(0, parameters));
        FixedStep_nodes.add(new GenerateInitialPointNode(1));
        FixedStep_nodes.add(new TerminationViaMaxTime<>(2, maxTime, "startTime"));
        FixedStep_nodes.add(new SampleNewPointNode_FixedStep(3));
        FixedStep_nodes.add(new SetBestNode(4));

        List<Tuple3<Integer, Option<Integer>, Integer>> FixedStep_transitionMatrix = new ArrayList<>();
        FixedStep_transitionMatrix.add(Tuple3.apply(0, Option.empty(), 1));
        FixedStep_transitionMatrix.add(Tuple3.apply(1, Option.empty(), 2));
        FixedStep_transitionMatrix.add(Tuple3.apply(2, Option.apply(0), 3));
        FixedStep_transitionMatrix.add(Tuple3.apply(2, Option.apply(1), 4));
        FixedStep_transitionMatrix.add(Tuple3.apply(3, Option.empty(), 2));

        return new Algorithm<>(
                JavaConverters.iterableAsScalaIterable(FixedStep_nodes).toSeq(),
                JavaConverters.iterableAsScalaIterable(FixedStep_transitionMatrix).toSeq());
    }
}